bitbots_head_mover/src/move_head.cpp

```diff
@@ -0,0 +1,679 @@
+#include <tf2_ros/transform_listener.h>
+#include <tf2_ros/buffer.h>
+#include <tf2_ros/transform_broadcaster.h>
+#include <tf2/convert.h>
+#include <bio_ik/bio_ik.h>
+#include <bio_ik_msgs/msg/ik_response.hpp>
+
+#include "head_parameters.hpp"
+
+#include <rclcpp_action/rclcpp_action.hpp>
+#include <bitbots_msgs/action/look_at.hpp>
+
+
+using std::placeholders::_1;
+using namespace std::chrono_literals;
+
+namespace move_head {
+  using LookAtGoal = bitbots_msgs::action::LookAt;
+  using LookAtGoalHandle = rclcpp_action::ServerGoalHandle<LookAtGoal>;
+
+
+class HeadMover {
+  std::shared_ptr<rclcpp::Node> node_;
+
+  //declare subscriber and publisher
+  rclcpp::Subscription<humanoid_league_msgs::msg::HeadMode>::SharedPtr head_mode_subscriber_;
+  rclcpp::Subscription<sensor_msgs::msg::JointState>::SharedPtr joint_state_subscriber_;
+  rclcpp::Subscription<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr ball_subscriber_;
+
+  rclcpp::Publisher<bitbots_msgs::msg::JointCommand>::SharedPtr position_publisher_;
+
+  // declare tf
+  std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
+  std::shared_ptr<tf2_ros::TransformBroadcaster> br_;
+  std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
+
+  //declare variables
+  humanoid_league_msgs::msg::HeadMode head_mode_;
+  sensor_msgs::msg::JointState current_joint_state_;
+  bitbots_msgs::msg::JointCommand pos_msg_;
+  double DEG_TO_RAD = M_PI / 180;
+  geometry_msgs::msg::PoseWithCovarianceStamped tf_precision_pose_;
+
+  //declare robot model and planning scene
+  robot_model_loader::RobotModelLoaderPtr loader_;
+  moveit::core::RobotModelPtr robot_model_;
+  moveit::core::RobotStatePtr robot_state_;
+  moveit::core::RobotStatePtr collision_state_;
+  planning_scene_monitor::PlanningSceneMonitorPtr planning_scene_monitor_;
+  planning_scene::PlanningScenePtr planning_scene_;
+
+  //declare world model variables
+  geometry_msgs::msg::PointStamped ball_;
+  geometry_msgs::msg::PointStamped ball_odom_;
+  geometry_msgs::msg::PointStamped ball_map_;
+  geometry_msgs::msg::PointStamped ball_teammate_;
+
+  //declare params
+  move_head::Params params_;
+
+  rclcpp::TimerBase::SharedPtr timer_;
+
+  std::vector<std::pair<double, double>> pattern_;
+  uint prev_head_mode_;
+
+  double threshold_;
+  int index_ = 0;
+  double pan_speed_;
+  double tilt_speed_;
+
+  // action server
+  rclcpp_action::Server<LookAtGoal>::SharedPtr action_server_;
+  bool action_running_ = false;
+
+
+ public:
+  HeadMover() {
+
+    node_ = std::make_shared<rclcpp::Node>("head_mover");
+
+    position_publisher_ = node_->create_publisher<bitbots_msgs::msg::JointCommand>("head_motor_goals", 10);
+    head_mode_subscriber_ =
+        node_->create_subscription<humanoid_league_msgs::msg::HeadMode>(
+            "head_mode",
+            10,
+            [this](const humanoid_league_msgs::msg::HeadMode::SharedPtr msg) { head_mode_callback(msg); }); // should be callback group 1
```

Comment:
why is this wrapped inside a lambda function?

Reply:
It is not wrong, but what is the reason?

Reply:
Like in the action server you could use `std::bind` imo.

Reply:
No real reason. I think for the action server, that is how they did it in the documentation. Should I rather make it uniform and use bind?

Reply:
I would prefer the usage of `std::bind`, but give it a test, as I am not 100% sure regarding the exact implementation details.